### PR TITLE
Feature/add catch errors middleware

### PR DESCRIPTION
**Pull request by tomchentw**
## Description

Add `catchError` middleware to validate `options` object. It may be very useful **in development**. Let me know what you think. Please note that the code is _not well organized_. Just want to get a first version and your feedback.

## TODOs

* [ ] Update ` CHANGES.md`
* [ ] Add `process.env.NODE_ENV` check
* [ ] Clean up the code to match the general style of this project.

## References

* https://developer.mozilla.org/en-US/docs/Web/API/GlobalFetch/fetch#Parameters

### CODE REVIEW

File: modules/index.js [hunk 1]
```diff
@@ -163,6 +163,59 @@ export const params = (object) => {
 }
 
 /**
+ * Check agains the final input and options to check if there's missing/misconfigured
+ * properties.
+ */
+export const catchErrors = () => {
+  function checkOptions(input, options) {
```
Comment (tomchentw) [inline modules/index.js]: Should I put this in the beginning of the file where `enhanceFetch`is?

File: modules/index.js [hunk 2]
```diff
@@ -163,6 +163,59 @@ export const params = (object) => {
...
+        }
+      },
+      {
+        name: 'body',
+        validate: (input, options) => {
+          const verb = (options.method || 'GET').toUpperCase()
+          if (verb === 'GET' || verb === 'HEAD') {
+            return
+          }
+          function isPojo(obj) {
```
Comment (tomchentw) [inline modules/index.js]: Should I use a npm version of this function? It's https://www.npmjs.com/package/is-pojo

File: modules/index.js [hunk 3]
```diff
@@ -163,6 +163,59 @@ export const params = (object) => {
...
+  function checkOptions(input, options) {
+    [
+      {
+        name: 'data',
+        validate: (input, options) => {
+          const verb = (options.method || 'GET').toUpperCase()
+          if (verb === 'GET' || verb === 'HEAD') {
+            return
+          }
+          throw new Error(`Did you try to use 'body: ${JSON.stringify(options.data)}'?`)
```
Comment (tomchentw) [inline modules/index.js]: Maybe use [`invariant`](https://www.npmjs.com/package/invariant)?

File: modules/index.js [hunk 4]
```diff
@@ -163,6 +163,59 @@ export const params = (object) => {
...
+    ].forEach(({ name, validate }) => {
+      if (name in options) {
+        validate(input, options)
+      }
+    })
+  }
+
+
+  return (fetch, input, options = {}) => {
+    checkOptions(input, options)
```
Comment (tomchentw) [inline modules/index.js]: Maybe add a `process.env.NODE_ENV` check?

**Comment by chentsulin** [comment]
@tomchentw maybe you want to submit this to the repo which against http-client debug: https://github.com/chentsulin/http-client-debug

**Comment by tomchentw** [comment]
Good to hear that! @chentsulin

**Pull request closed without merging**